### DISCUSSION TITLE
Access remaining rate limit

### DISCUSSION THREAD

**Discussion by abreumatheus** [joalla/discogs_client#21] [Ideas] [answered]
Hi,
A problem that i found on the official fork is that if i wanted to access the remaining limit i had to get it from a private var. Have you thought of exposing this value? Maybe even create a decorator, that automatically stop for a second if you reach 60 calls.

**Comment by alifhughes** [comment]
Hey @abreumatheus, as you've said - we did a quick fix when we first made this fork we exposed them as variables in the fetcher classes, [here](https://github.com/joalla/discogs_client/blob/master/discogs_client/fetchers.py#L64-L69), [here](https://github.com/joalla/discogs_client/blob/master/discogs_client/fetchers.py#L102-L107) and [here](https://github.com/joalla/discogs_client/blob/master/discogs_client/fetchers.py#L47-L52) - So I hope that helps as an intermediate solution. We were undecided whether or not to handle the throttling ourselves, there was a discussion around [who has the responsibilities of this in the original fork](https://github.com/discogs/discogs_client/issues/92#issuecomment-374327982). It seems that we're having more conversations around this again so I'll raise it with the other maintainer @JOJ0 and see what we think, and get back to you asap. Hope this is okay. Thanks

**Comment by JOJ0** [comment]
Hi @abreumatheus, thanks for mentioning this topic here. We are aware of it but haven't gotten around to discussing what we think is the right direction to go with it. BTW people hitting rate limits is one of the most reported topics on the Discogs forum too :-) @alifhughes and me now took some minutes to talk about it again and decided that we should try to find the people that originally filed PR's targeting rate-limit handling.

This one, originally submitted by @hiaselhans in June 2019 looks promising:
https://github.com/discogs/discogs_client/pull/102/commits/ff4036fec52d8e3050c03311bd7db8b2ad8e8536

@harababurel also submitted a PR to the original Discogs repo in April 2020, it makes use of the third party module "backoff" (https://github.com/litl/backoff):
https://github.com/discogs/discogs_client/pull/109/commits/1836fdc898ead72374a353011ec8df4f8d9a3272
As far as I understand this module is pretty easy to use, eg. a method is just decorated and backoff jumps in when the method throws an exception. Also another decorator can be used that jumps in on any condition. The drawback of using a third party module is just the fact that it is one ;-) We are pulling in another required python dependency...

I haven't tried both of the solutions myself and can't really tell which one is the better one but @alifhughes and me also discussed that maybe it should even be possible that the rate-limiting solution should be configurable on initialisation of the client class.

Give us your thoughts and if someone finds the time to try out one of those suggested PR's, it would help us a lot.

So as already said @hiaselhans and @abreumatheus, if you are still around, we are welcoming you to submit your PR's and let's work on them together :-) Thanks a lot in advance!

**Comment by hiaselhans** [comment]
Hi all!

It's already quite some time ago but it seems like i still have my fork.
It comes with a property ratelimit_enabled. I used it for work back then.

Let me know once you have decided and i could rebase and file the PR here.
You can also just copy the relevant parts if you like.

Have Fun!

**Comment by JOJ0** [comment]
Hi @hiaselhans thanks for getting back to us so quickly! It would be wonderful if you'd file the PR and we work from there. Probably there'll be slight changes necessary. This will help us a lot! Thanks in advance. Great!

**Comment by abreumatheus** [comment]
Hi guys, sorry for the delay on my side. I paticurlaly liked the @harababurel a little bit more, seems like the simpler it can get. Maybe the only modification would be expose a property like the `ratelimit_enabled ` idea from @hiaselhans, to make it optional, but defaulting to `True` .

**Comment by alifhughes** [comment]
> Hi guys, sorry for the delay on my side. I paticurlaly liked the @harababurel a little bit more, seems like the simpler it can get. Maybe the only modification would be expose a property like the `ratelimit_enabled ` idea from @hiaselhans, to make it optional, but defaulting to `True` .

I agree, I think @JOJ0 will too. I will get a PR open with this implementation anyway and we can discuss it further there, if we implement it and decide to go on a slightly different route, the PR will be the right way to capture that. I'll edit this question in the future with a link to the PR so that people will know where to go. Thanks for the good discussion guys, feel free to mark this as unanswered if we have more to say!

**Comment by JOJ0** [comment]
Actually I liked @hiaselhans more because it does not pull in another module, even though I think backoff seems to be a very well maintained package, but still: the best way would be if the user can choose betwee: no rate-limit-handling, backoff, builtin (hiaselhans) version. something like that. let's wait until @alifhughes and @hiaselhans do the PR's and then think about how to make it configurable. at least it would be good to get any of the two into the code.

**Reply by VinylVault** [reply]
@JOJ0 surely, rate-limit protection should just be enforced, no one wants to get 429's

**Reply by alifhughes** [reply]
@VinylVault I think that is making an assumption of our users of the client. _You could_ argue we could only be a layer over the API and not offer much more of our own functionality on top of it ([which was argued in a PR against the original repo](https://github.com/discogs/discogs_client/pull/89#issue-176067499)). However I think we're moving towards more than that. So it doesn't hurt for us to enable it by default and then **offer** the user of the client to not allow it if they wish. It is win win. I made a start to the development of this over the weekend but a bit short on time at the moment. Hope to have it done soon.

**Reply by VinylVault** [reply]
@alifhughes I agree that there are advantages and disadvantages to default ON, but to me, it's always better to cater to the learners / less experienced users of a language, than expecting that everyone can handle the error out of the box.

Part of the reason I would like to see that, is purely selfish, I use a specific project as my gateway project to learn a language, but doing that, enables me to see where there are potential issues with any module / helper, that other people may be struggling with, but don't want to raise an issue for the fear of looking like a dummy ... me I don't care about looking stupid, I know I'm learning :D

**Comment by alifhughes** [comment]
Just wanted to post an update that I haven't forgotten about this thread or the rate limiting, I have a **work in progress, draft PR** [here](https://github.com/joalla/discogs_client/pull/34) that I think could be the solution. Hope to get it complete and ready for review asap. 🙂

**Reply by VinylVault** [reply]
any news @alifhughes

**Comment by JOJ0** [comment, marked as answer]
As PR #34 will be merged sooner or later I will mark this question as answered :-)

**Reply by alifhughes** [reply]
This has now been released in version 2.3.10 - link to download on PyPI [here](https://pypi.org/project/python3-discogs-client/2.3.10/)
